### PR TITLE
Change the ace editor to monaco.

### PR DESCRIPTION
**Pull request by knightburton**
- Remove the ace editor.
- Init the monaco editor at the start of the page.
- Modify the file based function in the session module.
- Modify the file swith logic based on the monaco model structure.
- Modify the line highlighter: execute line and exception line.
- Modify the editor event listeners: mouseDown, modelChange.
- Modify the available breakpoint highlighter function.
- Modify the breakpoint insert and delete function.
- Modify the global theme settings to adapt to the monaco.
- Modify the editor related settings in settings menu to adapt to the monaco.
- Modify the load, reload functions in the console to adapt to the monaco.
- Modify the completer module to adapt to the monaco.
- Modify a few layout related code and css to display the monaco properly.
- Add a new iotjs function list to the project which is compatible with monaco.

IoT.jsCode-DCO-1.0-Signed-off-by: Imre Kiss kissi.szeged@partner.samsung.com

### CODE REVIEW

**Comment by knightburton** [comment]
- [x] Remove the ace editor.
- [x] Init the monaco editor.
- [x] Open file content.
- [x] Remove file content.
- [x] Switch between contents.
- [x] Update file related functions in the Session module to adapt to the new Model based structure.
- [x] Highlight the currently executed line by the engine in the editor workspace.
- [x] Highlight the exceptional line in the editor workspace.
- [x] Add mosueClick logic to the monaco gutter line numbers to handle the breakpoint insert/delete.
- [x] Add contentChange logic to the monaco workspace to handle the available breakpoint marks.
- [x] Show the available breakpoint lines in the monaco gutter.
- [x] Add Breakpoint insert and delete logic to the available lines.
- [ ] Adapt the source loader and reloader in the console panel to the monaco.
- [ ] Create a new autocompleter to display the available IoT.js functions in the intellisense context menu.
- [ ] Adapt the global theme settings to the monaco.
- [ ] Adapt the editor settings to the monaco.
- [ ] Customize and check the style of the editor.

**Comment by knightburton** [comment]
Well, this is a really big pull request, but I did not want to split this into small patches.

Note-1: The iotjs function json file is generated from the iotjs api documentation files. This generator script is not publish ready yet.
Note-2: The monaco's intelliSense has so many possibilities, so the current state of the Completer module is temporary.

**Comment by yichoi** [comment]
@daeyeon

File: src/app/surface.js [hunk 1]
```diff
@@ -796,14 +797,23 @@ export default class Surface {
     $('#editor').height($('#editor-wrapper').height() - height - 3);
   }
 
+  getEditorContainerDimensions() {
+    const height = ($('#file-tabs').children().length) ? $('#file-tabs').height() : this._editor.minTabHeight;
+
+    return {
+      width: $('#editor-wrapper').width() - 2,
```
Comment (daeyeon) [inline src/app/surface.js]: Can we avoid hardcoded offsets here?
I can find the similar offset in the above code, `.height($('#editor-wrapper').height() - height - 3);`

File: src/app/client-connection.js [hunk 2]
```diff
@@ -357,10 +357,10 @@ function onmessage(event) {
           this._exceptionData = null;
         }
       } else {
-        // Highlight the execute line in the correct session.
-        if (sID !== undefined && sID === this._session.getActiveID()) {
+        // Hightlight the execute line in the correct session.
```
Comment (daeyeon) [inline src/app/client-connection.js]: highlight

File: src/app/client-debugger.js [hunk 3]
```diff
@@ -712,25 +712,25 @@ export default class DebuggerClient {
 
     this.setEngineMode(ENGINE_MODE.RUN);
 
-    let sid = this._session.getUploadList()[0];
+    let fid = this._session.getUploadList()[0];
```
[the remark this answers is not in the document]
Comment (knightburton) [reply src/app/client-debugger.js]: Sorry, my wording was not the best in the doc. So, the `getUploadList` returns the array of file ids which were selected for upload. I fixed the doc, but I think the api name is fine.

File: src/app/surface.js (hunk 1 above)
Comment (knightburton) [reply src/app/surface.js]: Yes, of course, I forgot about this.

File: src/app/main.js [hunk 4]
```diff
@@ -187,6 +204,9 @@ export default function App() {
         if (surface.getPanelProperty('chart.active')) {
           surface.initChartPanel(chart);
         }
+
+        // Adapt the editor dimensions.
+        env.editor.layout(surface.getEditorContainerDimensions());
```
Comment (daeyeon) [inline src/app/main.js]: I found a bug. `env.editor.layout(surface.getEditorContainerDimensions());` need to be called upon `.extra-sidenav-toggle` also if there is no active panel after toggling the last one.

Comment (daeyeon) [reply src/app/main.js]: I'm fine with your opinion.

**Comment by yichoi** [comment]
LGTM

**Pull request merged**